### DISCUSSION TITLE
Restricting parent of polynomial to a subring

### DISCUSSION THREAD

**Discussion by oskarhenriksson** [oscar-system/Oscar.jl#3074] [Q&A] [answered]
Suppose I have a polynomial $f\in k[x_1,\ldots,x_n]$ such that all nonzero terms involve only the variables $x_1,\ldots,x_k$ for $k < n$. 

Is there a way create a new polynomial $\hat{f}$ with the same terms, but whose parent ring instead is $k[x_1,\ldots,x_k]$? 

(This would be analogous to how `to_univariate` function works in the $k=1$ case.)

**Example:**
```
R, (x,y,z) = polynomial_ring(QQ,["x","y","z"])
f = x^2+y^2+1
```

**Comment by fieker** [comment]
On Wed, Dec 06, 2023 at 11:08:27AM -0800, Oskar Henriksson wrote:
> Suppose I have a polynomial $f\in k[x,y,z]$ such that all nonzero terms involve only the variables $x$ and $y$. 
> 
>  Is there a way to modify $f$ so that its parent ring instead becomes $k[x,y]$? (This would be a bit like how the `to_univariate` function works.)
> 
> **Example:**
> ```
> R, (x,y,z) = polynomial_ring(QQ,["x","y","z"])
> f = x^2+y^2+1
> ```
> 
Well....

S, (x,y) = polynomial_ring(QQ,["x","y","z"])
z = Oscar.MPolyHom_vars{typeof(S), typeof(R)}(S, R, type = :names)
preimage(z, f)


could be improved....


z is a map from S to R mapping vaiables to vaiables with the same name.
The actual mapping is efficient.

**Reply by oskarhenriksson** [reply]
Hm, if I run the suggested commands:

```
R, (x,y,z) = polynomial_ring(QQ,["x","y","z"])
f = x^2+y^2+1
S, (x,y) = polynomial_ring(QQ,["x","y","z"])
z = Oscar.MPolyHom_vars{typeof(S), typeof(R)}(S, R, type = :names)
preimage(z, f)
```

the resulting preimage will still have a polynomial ring in *three* variables as its parent. Is there a way to make the parent a polynomial ring with *two* variables?

Intuitively, the following would make sense to me, but it gives an error message (BoundsError: attempt to access 2-element Vector{Int64} at index [3]).

```
R, (x,y,z) = polynomial_ring(QQ,["x","y","z"])
f = x^2+y^2+1
S, (x,y) = polynomial_ring(QQ,["x","y"])
phi = Oscar.MPolyHom_vars{typeof(S), typeof(R)}(S, R, type = :names)
g = preimage(phi, f)
```

**Reply by wdecker** [reply, marked as answer]
R, (x,y,z) = polynomial_ring(QQ,["x","y","z"])
f = x^2+y^2+1
S, (x,y) = polynomial_ring(QQ,["x","y"])
phi = hom(R, S, [x, y, 0])
f =phi(f)
